File: HRForz-API-Node/src/shared/schemas/response.ts
export interface ApiResponse<T> {
  message: string;
  response: T | null;
  code: string;
}

export interface PaginationMeta {
  page: number;
  pageSize: number;
  totalRecords: number;
  totalPages: number;
  hasNext: boolean;
  hasPrev: boolean;
}

export interface PaginatedData<T> {
  data: T[];
  meta: PaginationMeta;
}

export interface PaginatedResponse<T> {
  message: string;
  response: PaginatedData<T>;
  code: string;
}

export class ApiResponseBuilder {
  static ok<T>(data: T, message: string = 'Success'): ApiResponse<T> {
    return {
      message,
      response: data,
      code: 'SUCCESS',
    };
  }

  static created<T>(data: T, message: string): ApiResponse<T> {
    return {
      message,
      response: data,
      code: 'CREATED',
    };
  }

  static noContent(message: string = 'Deleted successfully'): ApiResponse<null> {
    return {
      message,
      response: null,
      code: 'SUCCESS',
    };
  }

  static paginated<T>(
    data: T[],
    total: number,
    page: number,
    pageSize: number,
    message: string = 'Records fetched'
  ): PaginatedResponse<T> {
    const totalPages = Math.max(1, Math.ceil(total / pageSize));
    return {
      message,
      response: {
        data,
        meta: {
          page,
          pageSize,
          totalRecords: total,
          totalPages,
          hasNext: page < totalPages,
          hasPrev: page > 1,
        },
      },
      code: 'SUCCESS',
    };
  }
}

File: HRForz-API/app/shared/schemas/response.py
from typing import Any, Generic, TypeVar

from pydantic import BaseModel

T = TypeVar("T")


class ApiResponse(BaseModel, Generic[T]):
    """Every single API endpoint returns this shape."""

    message: str
    response: T | None = None
    code: str = "SUCCESS"

    @classmethod
    def ok(cls, data: T, message: str = "Success") -> "ApiResponse[T]":
        return cls(message=message, response=data, code="SUCCESS")

    @classmethod
    def created(cls, data: T, message: str) -> "ApiResponse[T]":
        return cls(message=message, response=data, code="CREATED")

    @classmethod
    def no_content(cls, message: str = "Deleted successfully") -> "ApiResponse[None]":
        return cls(message=message, response=None, code="SUCCESS")


class PaginatedResponse(BaseModel, Generic[T]):
    message: str
    response: dict[str, Any]
    code: str = "SUCCESS"

    @classmethod
    def ok(
        cls,
        data: list[T],
        total: int,
        page: int,
        page_size: int,
        message: str = "Records fetched",
    ) -> "PaginatedResponse[T]":
        total_pages = max(1, -(-total // page_size))
        return cls(
            message=message,
            response={
                "data": data,
                "meta": {
                    "page": page,
                    "pageSize": page_size,
                    "totalRecords": total,
                    "totalPages": total_pages,
                    "hasNext": page < total_pages,
                    "hasPrev": page > 1,
                },
            },
            code="SUCCESS",
        )
